onert-micro/onert-micro/src/core/train/OMCheckpointSaver.cpp

```diff
@@ -0,0 +1,433 @@
+
+        // Obtain m and v data from train storage
+        uint8_t *m_data = adam_opt->getExponentAvgDataByTensorIndex(input_tensor_index);
+        assert(m_data != nullptr);
+        uint8_t *v_data = adam_opt->getExponentAvgSquaresDataByTensorIndex(input_tensor_index);
+        assert(v_data != nullptr);
+
+        // Write m data
+        cur_ptr = data.data() + m_buffer_offset;
+        std::memcpy(cur_ptr, m_data, buffer_size);
+        // Write v data
+        cur_ptr = data.data() + v_buffer_offset;
+        std::memcpy(cur_ptr, v_data, buffer_size);
+
+        // Move m and v buffers offsets
+        m_buffer_offset += buffer_size;
+        v_buffer_offset += buffer_size;
+      }
+    }
+
+    // Set pointer to the m offset field start
+    char *m_offset_start = data.data() + m_offset;
+    // Set pointer to the v offset field start
+    char *v_offset_start = data.data() + v_offset;
+    // Write m and v offsets
+    for (uint32_t i = 0; i < tensors_size; ++i)
+    {
+      // Save m buffer offset
+      uint32_t cur_m_offset = m_offsets[i];
+      std::memcpy(m_offset_start, &cur_m_offset, sizeof(cur_m_offset));
+      m_offset_start += 4;
+
+      // Save v buffer offset
+      uint32_t cur_v_offset = v_offsets[i];
+      std::memcpy(v_offset_start, &cur_v_offset, sizeof(cur_v_offset));
+      v_offset_start += 4;
+    }
+  }
+  else
+  {
+    // Note: offset = 0 - means there are no such buffers
+    // Move pointer to m offset field
+    cur_ptr = data.data() + 4;
+    // Save offset for the m offset field
+    uint32_t m_offset = 0;
+    std::memcpy(cur_ptr, &m_offset, sizeof(m_offset));
+
+    // Move pointer to the v offset field
+    cur_ptr += 4;
+    // Save offset for the v offset field
+    uint32_t v_offset = 0;
+    std::memcpy(cur_ptr, &v_offset, sizeof(v_offset));
+  }
+
+  // Move cur_ptr to the start of the offsets field
+  cur_ptr = data.data() + 20;
+  // Write offsets
+  for (uint32_t i = 0; i < tensors_size; ++i)
+  {
+    uint32_t offset = offsets[i];
+    std::memcpy(cur_ptr, &offset, sizeof(offset));
+    cur_ptr += 4;
+  }
+
+  // Save other parameters offset: 20 initial bytes + tensors_size * 4 bytes for buffer offsets +
+  // buffer size
+  uint32_t other_parameters_offset = 20 + tensors_size * 4 + acc_buffer_size;
+  // Adam case need add two more acc_buffer_size
+  if (config.training_context.optimizer == ADAM and not train_storage.getAdam()->isReset())
+  {
+    other_parameters_offset += acc_buffer_size * 2;
+    other_parameters_offset += tensors_size * 4 * 2;
+  }
+
+  // Write this offset
+  cur_ptr = data.data() + 12;
+  std::memcpy(cur_ptr, &other_parameters_offset, sizeof(other_parameters_offset));
+
+  // Move pointer to other parameters offset
+  cur_ptr = data.data() + other_parameters_offset;
+
+  // Write current step
+  std::memcpy(cur_ptr, &config.training_context.num_step, sizeof(config.training_context.num_step));
+
+  cur_ptr += 4;
+  // Write current epoch
```

Comment:
I am not sure that epoch is required for checkpoint. But just keep it as it is, now.

Reply:
Okay, I will keep it :) I think it can be helpful information to get some statistic from training. And this information costs only 4 bytes